### PR TITLE
agents: click-into-subagent navigation + auto-compaction (exploration never overflows)

### PR DESCRIPTION
**Pull request by agjs**
Follow-up to #78 (model-driven delegation). Two threads: finish the **navigable** half of the live agent tree, and make subagent exploration **robust** — a subagent must never fail just because its context grew.

### 1. Editor-mode ↑/↓ agent navigation (`9d67daf`)
The arrow-key "click into an agent and watch it" navigation shipped in #78 only worked in the readline input path; the default multi-line editor owns stdin exclusively and consumed ↑/↓ for cursor/history. Now, on an **empty** input row, ↑/↓ move the detail-pane focus between running subagents; while you're composing a mid-run steering message the arrows still edit text (never hijacked). Parity with the readline path via a new `onNavigateTree` editor hook.

### 2. Detail-pane label-echo fix (`22f3972`)
`renderEvent` had no case for the `agent_*` lifecycle kinds, so they fell through to the default branch and their text got routed into the agent's **own** detail buffer — printing the label 2–3 times under the tree row. Lifecycle events are structural (the tree renders them); when a tree observer is attached their text is no longer routed. Headless/one-shot keeps the linear-log line.

### 3. Subagent auto-compaction — the real robustness fix (`e895365`)
A read-only subagent (`AgentRunner`) composes the turn loop directly and never used `Session`, so it inherited **none** of the main loop's auto-compaction. History grew unbounded (every re-read re-appends the whole file) until the request overflowed the model window and the endpoint 400'd — failing exploration for no reason other than length.

**Confirmed live** on DeepSeek-V4-Flash (see the eval harness): `prompt 114689 + reserved output 16384 = 131073 > 131072 window`.

Fix mirrors `Session.compact`:
- **Proactive** — before a turn, if the previous prompt crossed 0.8 of the *effective* window (`window − reserved max_tokens`; that reserve is why the baseline 400'd at only 78.7% of the full window), summarize into `[system, summary]` and continue. The summarizer input is itself bounded so it can't overflow.
- **Reactive** — if one large read still jumps past the limit in a single turn, catch the context-overflow 400, compact, and retry once. Non-overflow errors still surface.
- Threads the session's context window into every subagent.

### 4. Eval harness (`scripts/agents-concurrency-eval.ts`)
Reusable harness that fans out `explore` subagents at several concurrency caps against the live endpoint and reports wall-clock, aggregate tok/s, peak prompt tokens, per-agent compaction count, and verbatim failures. This is what measured the 400 and the concurrency behaviour (cap>1 is currently a net negative on the MTP endpoint — the subagent value is context isolation, not wall-clock; default `agents.concurrency: 1` stays correct).

## Tests
- `agent-compaction` — proactive / no-op / reactive-recovery / non-overflow-not-swallowed
- `editor-controller` — ↑/↓ forwarded on empty buffer, NOT while composing, declined keys fall through to history
- `logging-lifecycle` — lifecycle text not routed to the agent sink under a tree; real tool/token output still routed; headless linear-log preserved
- `scripts/e2e-spawn-agent-pty.py` — real-PTY: drives ↑/↓ in editor mode and asserts the detail pane moves BACK to the earlier agent (impossible via auto-follow), then forward again

Full `bun run validate` green: **2023 pass / 0 fail**, all PTY suites pass, docs build clean.

### CODE REVIEW

File: packages/core/src/agent/agent-runner.ts [hunk 1]
```diff
@@ -427,4 +541,74 @@ export class AgentRunner {
...
+  ): Promise<IModelResponse> {
+    const window = opts.contextWindow ?? 0;
+    const provider = opts.provider;
+    // vLLM counts prompt + reserved OUTPUT tokens against the window, so the
+    // real ceiling for the PROMPT is `window - maxTokens`. Trigger compaction
+    // against that effective budget — otherwise a request can 400 while the
+    // prompt alone is still under `window` (observed: 114689 prompt + 16384 out
+    // = 131073 > 131072). The reactive path below is the backstop for a single
+    // read that jumps past the threshold in one turn.
+    const effectiveWindow = Math.max(1, window - PROVIDER_LIMITS.maxTokens);
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/agent/agent-runner.ts]: Fixed in 4d88700. The output reserve is now capped at half the window: \`reserve = min(PROVIDER_LIMITS.maxTokens, floor(window * 0.5))\`, so a small model can never collapse \`effectiveWindow\` toward 1 and compact every turn. Added a regression test (`a small context window does NOT trigger compaction every turn`, contextWindow=8000).

File: packages/core/src/agent/agent-runner.ts [hunk 2]
```diff
@@ -22,14 +22,115 @@ import {
...
+/** A vLLM/OpenAI-style context-overflow rejection ("This model's maximum context
+ *  length is N tokens. However, you requested M…"). Matched so it can be RECOVERED
+ *  (compact + retry) instead of failing the agent. */
+function isContextOverflow(err: unknown): boolean {
+  const msg = err instanceof Error ? err.message : String(err);
+
+  return /maximum context length|context length|context window|too many tokens|reduce the length|too long/iu.test(
+    msg
+  );
+}
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/agent/agent-runner.ts]: Fixed in 4d88700. Added an \`errorText(err)\` helper that pulls the message from non-Error throws too — a plain object with \`message\`, or a nested \`error.message\` — using the existing \`isRecord\` guard (no casts, per house rules). \`isContextOverflow\` now runs the regex over that. Tests cover the nested-object case.

File: packages/core/src/agent/agent-runner.ts [hunk 3]
```diff
@@ -22,14 +22,115 @@ import {
...
+
+    if (msg === undefined) {
+      continue;
+    }
+
+    const text = line(msg);
+
+    if (used + text.length > maxChars) {
+      break;
+    }
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/agent/agent-runner.ts]: Fixed in 4d88700. \`buildBoundedTranscript\` now fills newest-first and ALWAYS includes at least the most recent message (head-truncating it to fit if it alone exceeds the budget), so the overflow-causing context is never dropped. Test: `always includes the newest message, truncating it if it alone exceeds the budget`.

File: packages/core/src/agent/agent-runner.ts [hunk 4]
```diff
@@ -22,14 +22,115 @@ import {
...
+
+    tail.unshift(text);
+    used += text.length;
+  }
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/agent/agent-runner.ts]: Fixed in 4d88700. Two changes guarantee the bound: (1) each message is individually capped to the budget (so the first/only message can no longer exceed \`maxChars\`), and (2) \`TRANSCRIPT_MARKER_RESERVE\` (64 chars) is subtracted up front to leave room for the elision marker + join separators. Test asserts \`out.length <= maxChars\` for both the oversized-single-message and many-messages cases.

File: packages/core/src/agent/agent-runner.ts [hunk 5]
```diff
@@ -427,4 +541,74 @@ export class AgentRunner {
...
+    const window = opts.contextWindow ?? 0;
+    const provider = opts.provider;
+    // vLLM counts prompt + reserved OUTPUT tokens against the window, so the
+    // real ceiling for the PROMPT is `window - maxTokens`. Trigger compaction
+    // against that effective budget — otherwise a request can 400 while the
+    // prompt alone is still under `window` (observed: 114689 prompt + 16384 out
+    // = 131073 > 131072). The reactive path below is the backstop for a single
+    // read that jumps past the threshold in one turn.
+    const effectiveWindow = Math.max(1, window - PROVIDER_LIMITS.maxTokens);
+    const at = AGENT_LIMITS.autoCompactAt;
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/agent/agent-runner.ts:683]: Good catch. Mitigated rather than fully parameterized: the reserve is now \`min(PROVIDER_LIMITS.maxTokens, floor(window*0.5))\`, and the 0.8 fraction leaves margin, so a model with a HIGHER configured maxTokens is caught by the reactive path (a context-overflow 400 → compact + retry once) which is the backstop for exactly this. Threading the per-model \`entry.maxTokens\` through is a clean follow-up if we see it trigger in practice.

File: packages/core/scripts/agents-concurrency-eval.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        );
+        const agentStart = performance.now();
+        const result = await new AgentRunner(explore).run({
+          provider: counter,
+          cwd,
+          parentTaskId: `cap${cap}-${i}`,
+          task: task.prompt,
+          report: silent,
+          policyMode: "bypassPermissions",
+        });
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/scripts/agents-concurrency-eval.ts:144]: Fixed in 4d88700 — this was the critical one. \`runOneCap\` now takes the resolved \`contextWindow\` and passes it into \`AgentRunner.run({... contextWindow})\`, so auto-compaction is actually exercised. The re-run (in progress) now reports \`compactions=N\` per agent instead of a flat 0.

**Comment by agjs** [comment]
Addressed all review feedback in 4d88700 (all threads replied + resolved):

- **Gemini (high) — small-window infinite compaction:** output reserve capped at ½ the window; regression test added.
- **Gemini/Copilot — buildBoundedTranscript:** now newest-first, always includes the most recent message (truncated if huge), and reserves marker/separator space so the transcript truly stays within `maxChars`.
- **Gemini — isContextOverflow on non-Error:** `errorText()` extracts `message` / nested `error.message` via `isRecord` (no casts).
- **Copilot (critical) — eval didn't pass contextWindow:** threaded `contextWindow` into `AgentRunner.run`, so compaction is actually exercised (`compactions=N` per agent now).
- **Copilot — configurable maxTokens reserve:** mitigated by the ½-window cap + 0.8 margin; the reactive overflow-recover path is the backstop.

Tests: 4 → 11 in `agent-compaction` (covering every edge case raised). Full `bun run validate` green.

Re-running the live eval now (MTP off + retuned endpoint) to post the concurrency-scaling + zero-failure numbers.

**Comment by agjs** [comment]
## Live eval results (DeepSeek-V4-Flash, MTP removed + retuned: max_num_seqs=16, batched_tokens=32768)

4 `explore` subagents fanned out across subsystems, at concurrency caps 1/2/4:

| cap | wall | agg tok/s | failed | engine |
|----|------|-----------|--------|--------|
| 1 | 591.6s | 19.3 | 0/4 | ok |
| 2 | 422.3s | 22.9 | 0/4 | ok |
| 4 | 495.9s | 18.6 | 0/4 | **ok (no crash)** |

### Auto-compaction proven live
```
cap 4:  ✓ loop  turns=9  peakPrompt=107283  compactions=1
```
The `loop` agent grew to **107,283** prompt tokens, **auto-compacted once**, and finished ✓. In the pre-fix baseline the same agent hit 114,689 → `400 maximum context length` → hard failure. This is the exact case the PR fixes, recovering live against the real model. **0/4 failed at every cap.**

### Concurrency (MTP off — the retune hermes applied)
Before (MTP on), cap 2 was *slower* than cap 1 (597s vs 449s) and cap 4 **crashed the vllm engine** (500 EngineCore). After removing MTP:
- cap 1→2: **591.6s → 422.3s**, tok/s **19.3 → 22.9** — concurrency now genuinely helps.
- cap 4 no longer crashes; it's slower than cap 2 here, partly confounded by an unrelated build I launched mid-run. cap 2 looks like the sweet spot for long-context explore agents.

**Takeaway:** shipped default `agents.concurrency: 1` stays correct (context-isolation is the guaranteed win); cap 2–4 is now *safe* (no crash) and a modest wall-clock win, where before it was destructive. The exploration-never-fails goal is met.

**Comment by agjs** [comment]
### Follow-up: proving + improving the subagent value

**A/B eval (delegation ON vs OFF, live DeepSeek-V4-Flash, n=1)** — the value is *grounding*:
- **OFF**: answered in 1 turn, peak 1804 tok — and **confabulated files that don't exist** (`src/gate.ts`, `src/agent-scheduler.ts`), with invented line numbers + code.
- **ON**: spawned subagents that actually read the code → **real** citations (`packages/core/src/gate/core-gate.ts:23`, `agent/agent-scheduler.ts:38`).
- Caveat: n=1; the isolation metric (orch peak) was *higher* for ON only because OFF punted without working. Needs repeats + a forced-inline arm.

**New commits on this branch:**
1. `loop:` **phase-2-skip fix** — staged-build shipped a hollow app (types-only phase 1 passes the gate → phase 2 skipped). Now skips only when real implementation exists. +regression test.
2. `agents:` **structured subagent output** — `agent_result` is now `{ summary, findings:[{detail, source, confidence}] }`; all built-ins run `outputMode:structured`, forcing cited evidence on every point.
3. `eval:` **TSFORGE_NO_DELEGATION kill-switch + delegation A/B harness** (config-agnostic, re-runnable for the MTP×delegation grid).
4. `eval:` **headless-build** — wire delegation + detect the real context window (was hardcoded 262144, mis-calibrating compaction).

Full `bun run validate` green: 2038 pass / 0 fail, all PTY suites (spawn-agent 7/7 with structured subagents).

File: packages/core/src/agent/agent-runner.ts [hunk 7]
```diff
@@ -22,14 +22,155 @@ import {
...
+    const text = capped(msg);
+
+    // Always keep at least the newest (picked is empty on the first iteration).
+    if (picked.length > 0 && used + text.length > budget) {
+      break;
+    }
+
+    picked.unshift(text);
+    used += text.length;
+    firstIdx = i;
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/agent/agent-runner.ts:135]: Fixed. buildBoundedTranscript now counts a `SEP` ("\n\n") per included message in the running total (both the break check and `used += text.length + SEP`), and the elision marker + its separator are covered by TRANSCRIPT_MARKER_RESERVE. Test added: 200 one-char messages stay within a 300-char budget (would blow past without separator accounting).

File: packages/core/src/agent/agent-runner.ts [hunk 8]
```diff
@@ -427,4 +639,80 @@ export class AgentRunner {
...
+    try {
+      res = await provider.complete(ctx.messages, callOpts);
+    } catch (err) {
+      if (
+        opts.signal?.aborted === true ||
+        !isContextOverflow(err) ||
+        ctx.messages.length <= 2
+      ) {
+        throw err;
+      }
```
[the remark this answers is not in the document]
Comment (agjs) [reply packages/core/src/agent/agent-runner.ts:718]: Fixed. Added FALLBACK_COMPACT_CHARS (48000 ≈ 12k tokens) — when the window is unknown (window ≤ 0), compactAgentMessages uses that fixed bound instead of maxChars=0 (unbounded), so a reactive-recovery compaction can never itself overflow. Test added: reactive overflow recovery succeeds with contextWindow omitted.
